packages/common-utils/src/renderChartConfig.ts

```diff
@@ -734,23 +734,50 @@ function renderLimit(
 // CTE (Common Table Expressions) isn't exported at this time. It's only used internally
 // for metric SQL generation.
 type ChartConfigWithOptDateRangeEx = ChartConfigWithOptDateRange & {
-  with?: { name: string; sql: ChSql }[];
   includedDataInterval?: string;
 };
 
-function renderWith(
+/**
+ * Type guard to check if an object is a ChSql instance
+ * @param obj The object to check
+ * @returns True if the object is a ChSql instance
+ */
+function isChSqlInstance(obj: any): obj is ChSql {
+  return (
+    obj != null &&
+    typeof obj === 'object' &&
+    'sql' in obj &&
+    'params' in obj &&
+    typeof obj.sql === 'string' &&
+    typeof obj.params === 'object'
+  );
+}
+
+async function renderWith(
   chartConfig: ChartConfigWithOptDateRangeEx,
   metadata: Metadata,
-): ChSql | undefined {
+): Promise<ChSql | undefined> {
   const { with: withClauses } = chartConfig;
-  if (withClauses) {
-    return concatChSql(
-      ',',
-      withClauses.map(clause => chSql`${clause.name} AS (${clause.sql})`),
-    );
-  }
-
-  return undefined;
+  const renderedClauses = withClauses?.map(async withClause => {
+    const { name, sql } = withClause;
+    if (typeof sql === 'string') {
+      return chSql`${name} AS (${chSql`${{ UNSAFE_RAW_SQL: sql }}`})`;
+    } else if (isChSqlInstance(sql)) {
+      return chSql`${name} AS (${sql})`;
+    } else {
+      return chSql`${name} AS (${await renderChartConfig(
```

Comment:
better to verify if sql is a valid chart config and throw if nothing matches

Reply:
Broke out the chart config block into an else-if and now else raises an exception.